### ISSUE TITLE
Does `upsert` cannot work in `$transaction`

### ISSUE THREAD

**Issue by shtse8** [prisma/prisma#14686]
### Bug description

Does `upsert` cannot work in `$transaction`? as it returns ` Prisma.Prisma__DataClient<Data>` instead of `PrismaPromise`.
I tried to use `upsert` in transaction, but the executed `upsert` before error didn't roll back.

it doesn't work in `$transaction([])`, but it works well in interactive transaction `$transaction(prisma => {})`.

the data doesn't roll back.
```ts
await prisma.$transaction([
      prisma.data.update({ where: { id }, data }),
      prisma.$executeRaw`TRUNCATE TABLE Words`,  // non exisiting table to force throw error
    ])
```

the data rolled back
```ts
await prisma.$transaction(async prisma => {
      await prisma.data.update({ where: { id }, data })
      await prisma.$executeRaw`TRUNCATE TABLE Words`  // non exisiting table to force throw error
    })
```

### How to reproduce

as above.

### Prisma information

as above.

### Environment & setup

- OS:  Windows 11, Ubuntu 22.4
- Database: Mysql 8
- Node.js version: 18.6

### Prisma Version

prisma                  : 4.1.1
@prisma/client          : 4.1.1
Current platform        : debian-openssl-3.0.x
Query Engine (Node-API) : libquery-engine 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Migration Engine        : migration-engine-cli 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules/@prisma/engines/migration-engine-debian-openssl-3.0.x)
Introspection Engine    : introspection-core 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-3.0.x)
Format Binary           : prisma-fmt 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-3.0.x)
Default Engines Hash    : 8d8414deb360336e4698a65aa45a1fbaf1ce13d8
Studio                  : 0.469.0
Preview Features        : interactiveTransactions

**Comment by janpio** [comment]
Hey @shtse8, can you please describe your problem again? 

Your issue title mentioned `upsert` but then all the examples (both with sequential and interactive transactions) actually use `update`. Is this a typo in the title or in the example? Or does it apply to both?

I assume the second query `prisma.$executeRaw'TRUNCATE TABLE Words'` is just a random query to trigger a failing transaction, which should lead to a rollback of the previously executed `update`? And that is not happening for you?

**Comment by shtse8** [comment]
> Hey @shtse8, can you please describe your problem again?
> 
> Your issue title mentioned `upsert` but then all the examples (both with sequential and interactive transactions) actually use `update`. Is this a typo in the title or in the example? Or does it apply to both?
> 
> I assume the second query `prisma.$executeRaw'TRUNCATE TABLE Words'` is just a random query to trigger a failing transaction, which should lead to a rollback of the previously executed `update`? And that is not happening for you?

I need some time to figure out what I was asking two months ago. Will reply soon.

**Comment by janpio** [comment]
Of course. Did you have time to get to that @shtse8? With the current information we unfortunately will need to close the issue as it is not enough to understand what is going on and reproduce it. Thanks!

**Comment by shtse8** [comment]
> Of course. Did you have time to get to that @shtse8? With the current information we unfortunately will need to close the issue as it is not enough to understand what is going on and reproduce it. Thanks!

will reply you in 7 days. sorry for my late. I am working on an urgent project this month.

**Comment by janpio** [comment]
@shtse8 Should we assume that you just meant `update` but typed `upsert`?
